src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/populatefromnomis/PopulateFromNomis.kt

```diff
@@ -0,0 +1,68 @@
+package uk.gov.justice.digital.hmpps.personrecord.populatefromnomis
+
+import feign.FeignException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestMethod
+import org.springframework.web.bind.annotation.RestController
+import org.springframework.web.client.HttpClientErrorException
+import org.springframework.web.client.HttpServerErrorException
+import uk.gov.justice.digital.hmpps.personrecord.client.PageParams
+import uk.gov.justice.digital.hmpps.personrecord.client.PrisonServiceClient
+import uk.gov.justice.digital.hmpps.personrecord.client.PrisonerSearchClient
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.repository.PersonRepository
+import uk.gov.justice.digital.hmpps.personrecord.model.Person
+import uk.gov.justice.digital.hmpps.personrecord.service.RetryExecutor.runWithRetry
+
+private const val OK = "OK"
+
+private val retryables = listOf(HttpClientErrorException::class, HttpServerErrorException::class, FeignException.InternalServerError::class, FeignException.ServiceUnavailable::class)
+
+@RestController
+class PopulateFromNomis(
+  val prisonerSearchClient: PrisonerSearchClient,
+  val prisonServiceClient: PrisonServiceClient,
+  @Value("\${populate-from-nomis.page-size}") val pageSize: Int,
+  @Value("\${populate-from-nomis.retry.delay}") val delayMillis: Long,
+  @Value("\${populate-from-nomis.retry.times}") val retries: Int,
+  val repository: PersonRepository,
+) {
+
+  @RequestMapping(method = [RequestMethod.POST], value = ["/populatefromnomis"])
+  suspend fun populate(): String {
+    populatePages()
+    return OK
+  }
+
+  suspend fun populatePages() {
+    // TODO understand scoping here, how many threads? This dispatcher will give up to 64
+    // https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-i-o.html
+    CoroutineScope(Dispatchers.IO).launch {
+      // if this call fails we will just restart the process, no need to retry
+      val prisonerNumbers = prisonServiceClient.getPrisonerNumbers(PageParams(0, pageSize))!!
+      val totalPages = prisonerNumbers.totalPages
+      var numbers = prisonerNumbers.numbers
+
+      for (page in 1..totalPages) {
+        numbers.forEach {
+          runWithRetry(retryables, retries, delayMillis) {
+            val prisoner = prisonerSearchClient.getPrisoner(it)
```

Comment:
I know you only have a single thread for now, but it's still going to put a lot of load onto prisoner search. I wonder if it's worth guaging how bad that hit on prisoner search is going to be first - maybe just do the first few pages or something?

Reply:
limited to 4000 records for now